grammar.js

```diff
@@ -64,372 +21,468 @@ module.exports = grammar({
-            ),
-          ),
-        ),
-      ));
-    },
-    CharLiteral: $ => {
-      // In Gosu's syntax, char literals have the same syntax as
-      // single-character single-quoted string literals. Give char literals
-      // higher precedence to break the conflict.
-      return  prec(1, seq("'", choice($.EscapeSequence, $.any_character), "'"));
-    },
-    StringLiteral: $ => choice(seq("'", repeat(choice($.EscapeSequence, $.any_character)), "'"), seq('"', repeat(choice($.EscapeSequence, $.any_character)), '"')),
-    HexDigit: $ => choice($.Digit, /[A-F]/, /[a-f]/),
-    IntegerTypeSuffix: _ => choice("l", "L", "s", "S", "bi", "BI", "b", "B"),
-    // "Letter" is defined above
     Digit: _ => Digit,
-    // "NonZeroDigit" is unused in any rule
-    ZeroToSeven: _ => /[0-7]/,
-    Exponent: $ => seq(choice("e", "E"), optional(choice("+", "-")), $.Digit, repeat($.Digit)),
-    FloatTypeSuffix: _ => choice("f", "F", "d", "D", "bd", "BD"),
-    EscapeSequence: $ => choice(seq("\\", choice("v", "a", "b", "t", "n", "f", "r", '"', "'", "\\", "$", "<")), $.UnicodeEscape, $.OctalEscape),
-    OctalEscape: $ => choice(seq("\\", /0-3/, $.ZeroToSeven, $.ZeroToSeven), seq("\\", $.ZeroToSeven, $.ZeroToSeven), seq("\\", $.ZeroToSeven)),
-    UnicodeEscape: $ => seq("\\u", $.HexDigit, $.HexDigit, $.HexDigit, $.HexDigit),
-    any_character: _ => {
-      // This rule is not defined in Gosu's grammar but it is used. Assuming it's equivalent to the "." in common regex syntax.
-      return /./;
-    },
-    WS: _ => /\s/,
+    // HexDigit: $ => choice($.Digit, /[A-F]/, /[a-f]/),
+    // ZeroToSeven: _ => /[0-7]/,
+    // EscapeSequence: $ => choice(seq("\\", choice("v", "a", "b", "t", "n", "f", "r", '"', "'", "\\", "$", "<")), $.UnicodeEscape, $.OctalEscape),
+    // OctalEscape: $ => choice(seq("\\", /0-3/, $.ZeroToSeven, $.ZeroToSeven), seq("\\", $.ZeroToSeven, $.ZeroToSeven), seq("\\", $.ZeroToSeven)),
+    // UnicodeEscape: $ => seq("\\u", $.HexDigit, $.HexDigit, $.HexDigit, $.HexDigit),
     COMMENT: _ => /\/\*([^\*]|(\*[^\/]))*\*\//,
     LINE_COMMENT: _ => /\/\/[^\n\r]*/,
   }
 });
+
+// `optionalSeq(a, b, c, ...)` is
+// `choice(
+//   seq(a, optional(b), optional(c)),
+//   seq(optional(a), b, optional(c)),
+//   seq(optional(a), optional(b), c)
+//  )`
+//  This gives us a tree-sitter compatible representation of rules like
+//  `header = ["package" namespaceStatement] [usesStatementList] .`
+//  where every subrule in the sequence is optional.
+function optionalSeq(...rules) {
+  if (rules.length === 0) {
+    throw new Error("optionalSeq requires at least one rule.");
+  }
+  let optional_rules = rules.map(
+    (_, non_optional_rule_idx) => {
+      let all_but_one_rule_optional = rules.map(
+        (rule, rule_idx) =>
+          non_optional_rule_idx === rule_idx
+            ? rule
+            : optional(rule)
+      )
+      return seq(...all_but_one_rule_optional)
+    }
+  )
+  return choice(...optional_rules)
+}
```

Comment:
This way of dealing with sequences where everything is optional is cleaner to read, IMO, and was recommended by folks on the tree-sitter discord. I suspect it has also helped reduce some of the ambiguity, but not sure because the grammar is so simple at current :D.